src/porepy/utils/grid_utils.py

```diff
@@ -104,3 +106,255 @@ def star_shape_cell_centers(g: "pp.Grid", as_nan: bool = False) -> np.ndarray:
+    # large/small cells
+    radius = 0.5 * (max_distance + min_distance) + 1e-15
+    if np.max((max_distance - min_distance) / radius) >= 1e-10:
+        raise ValueError("Circumcenter not equidistant from all nodes.")
+
+    # Build connectivity arrays
+    fn = np.reshape(sd.face_nodes.tocsc().indices, (sd.dim, -1), order="F")
+    cf = np.reshape(sd.cell_faces.tocsc().indices, (sd.dim + 1, -1), order="F")
+
+    xn = sd.nodes
+    # Compute face cross products and areas.
+    v1 = xn[:, fn[1]] - xn[:, fn[0]]
+    v2 = xn[:, fn[2]] - xn[:, fn[0]]
+
+    face_cross = np.vstack(
+        (
+            v1[1] * v2[2] - v1[2] * v2[1],
+            v1[2] * v2[0] - v1[0] * v2[2],
+            v1[0] * v2[1] - v1[1] * v2[0],
+        )
+    )
+
+    face_cross_area = np.linalg.norm(face_cross, axis=0)
+
+    distances_from_faces = []
+    inside_cell = []
+
+    for ci in range(cf.shape[0]):
+
+        fi = cf[ci]
+
+        # Is the face cross vector pointing into the cell?
+        cross_points_into_cell = np.sign(
+            np.sum(
+                face_cross[:, fi] * (sd.cell_centers - sd.face_centers[:, fi]),
+                axis=0,
+            )
+        )
+        # The height of the cell, measured as the distance from the current
+        # face to the oposite node
+        height = 6 * sd.cell_volumes / face_cross_area[fi]
+        unit_vec = face_cross[:, fi] / face_cross_area[fi]
+        # Distance from the plane of the current face to the computed circumcenter
+        distances_from_faces.append(
+            np.sum(unit_vec * (center.T - sd.face_centers[:, fi]), axis=0) / height
+        )
+        # Check whether the circumcenter is in the half space defined by the face
+        # normal pointing into the cell. This is the case if the distance has the same
+        # sign as the cross product vector.
+        inside_cell.append(
+            np.logical_and(
+                cross_points_into_cell == np.sign(distances_from_faces[-1]),
+                cross_points_into_cell != 0,
+            )
+        )
+    # The circumcenter is inside the cell if it is in the half space of all faces.
+    all_inside = np.all(inside_cell, axis=0)
+    logger.debug(f"{all_inside.sum()} out of {sd.num_cells} cells have the circumcenter"
+                  +"inside the cell.")
+    # Compute the minimum distance from faces, normalized with cell height.
+    distances_from_faces = np.array(distances_from_faces)
+    min_distance = np.min(np.abs(distances_from_faces), axis=0)
+    # Cells where the circumcenter is inside and at a reasonable distance from faces
+    # can have their centers replaced.
+    replace = np.logical_and(all_inside, min_distance > 0.05)
+    new_centers = sd.cell_centers.copy()
+    new_centers[:, replace] = center.T[:, replace]
+
+    logger.info(
+        "Replaced %d out of %d cell centers.", int(replace.sum()), int(sd.num_cells)
+    )
+
+    fc = sd.cell_faces_as_dense()
```

Comment:
Is it correct that from here on out, all calculations are sanity checks? If so, and if they were copied from the original code in the runscript repository, we can consider dropping them (I did not trust myself when writing that code).

Reply:
Yes and yes. Let's consider it when discussing other points.